In [1]:
from keras.datasets import imdb

NUM_WORDS = 10000

(dev_X, dev_y) , (test_X, test_y) = imdb.load_data(num_words=NUM_WORDS,
                                                      skip_top=0,
                                                      maxlen=None)

Using TensorFlow backend.


The point of all this thing is to try and beat their score
http://nbviewer.jupyter.org/github/fchollet/deep-learning-with-python-notebooks/blob/master/3.5-classifying-movie-reviews.ipynb

In [2]:
dev_X.shape

(25000,)

In [3]:
dev_y.shape

(25000,)

In [4]:
import numpy as np

In [5]:
from keras.layers import Dense, Dropout, Embedding, LSTM
from keras.models import Sequential
from keras.callbacks import TensorBoard

First I will try sequental model with few dense layers without autoencoder

In [6]:
def vectorize_y(Y):
    return np.asarray(Y).astype('float32')

In [7]:
y = vectorize_y(dev_y)
y[:2]

array([ 1.,  0.], dtype=float32)

In [8]:
def one_hot_encode(X):
    seq = np.zeros((len(X),NUM_WORDS))
    for (i,val) in enumerate(X):
        seq[i][val] = 1
    return seq

In [9]:
X = one_hot_encode(dev_X)
X[:2]

array([[ 0.,  1.,  1., ...,  0.,  0.,  0.],
       [ 0.,  1.,  1., ...,  0.,  0.,  0.]])

In [120]:
model = Sequential([
    Dense(16, input_shape=(NUM_WORDS,), activation="tanh"),
    Dropout(0.5),
    Dense(64, activation="tanh"),
    Dense(1, activation="sigmoid")
])

In [121]:
model.compile(optimizer="adam", loss="binary_crossentropy", metrics=["accuracy"])

In [122]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_70 (Dense)             (None, 16)                160016    
_________________________________________________________________
dropout_22 (Dropout)         (None, 16)                0         
_________________________________________________________________
dense_71 (Dense)             (None, 64)                1088      
_________________________________________________________________
dense_72 (Dense)             (None, 1)                 65        
Total params: 161,169
Trainable params: 161,169
Non-trainable params: 0
_________________________________________________________________


In [123]:
model.fit(X, y, epochs=3, validation_split=0.2, batch_size=712, callbacks=[TensorBoard(log_dir="./logs")])

Train on 20000 samples, validate on 5000 samples
Epoch 1/3
20000/20000 [==============================] - 1s 45us/step - loss: 0.5426 - acc: 0.7641 - val_loss: 0.3646 - val_acc: 0.8626
Epoch 2/3
20000/20000 [==============================] - 1s 42us/step - loss: 0.2909 - acc: 0.8914 - val_loss: 0.2804 - val_acc: 0.8886
Epoch 3/3
20000/20000 [==============================] - 1s 42us/step - loss: 0.2105 - acc: 0.9242 - val_loss: 0.2724 - val_acc: 0.8942


In [109]:
X_test = one_hot_encode(test_X)
y_test = vectorize_y(test_y)

In [124]:
results = model.evaluate(X_test, y_test)
results

25000/25000 [==============================] - 2s 62us/step


[0.2873726367521286, 0.88404000000000005]

Yup it seems I have won this one [0.2873726367521286, 0.88404000000000005] :)

And now last but not least, lets check how simple long short term memory network handles the classification

In [9]:
def pad_seq(X):
    max_len_seq = max([len(x) for x in X])
    seq = []
    for x in X:
        x_clone = np.append(x, [np.zeros((max_len_seq - len(x)), dtype="int32") ])
        seq.append(x_clone)
    return (max_len_seq, np.array(seq))

In [10]:
max_len_seq, X = pad_seq(dev_X)
X[:2]

array([[   1,   14,   22, ...,    0,    0,    0],
       [   1,  194, 1153, ...,    0,    0,    0]])

In [11]:
model = Sequential([
    Embedding(max_len_seq, 128),
    LSTM(128),
    Dense(1, activation="sigmoid")
])

In [12]:
model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [13]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, None, 128)         319232    
_________________________________________________________________
lstm_1 (LSTM)                (None, 128)               131584    
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 129       
Total params: 450,945
Trainable params: 450,945
Non-trainable params: 0
_________________________________________________________________


In [14]:
model.fit(X, y, epochs=12, validation_split=0.2, batch_size=256)

Train on 20000 samples, validate on 5000 samples
Epoch 1/12
20000/20000 [==============================] - 193s 10ms/step - loss: 0.6934 - acc: 0.5014 - val_loss: 0.6931 - val_acc: 0.5062
Epoch 2/12
20000/20000 [==============================] - 190s 10ms/step - loss: 0.6933 - acc: 0.4955 - val_loss: 0.6933 - val_acc: 0.4938
Epoch 3/12
20000/20000 [==============================] - 185s 9ms/step - loss: 0.6933 - acc: 0.4971 - val_loss: 0.6932 - val_acc: 0.4938
Epoch 4/12
20000/20000 [==============================] - 184s 9ms/step - loss: 0.6932 - acc: 0.4971 - val_loss: 0.6934 - val_acc: 0.4938
Epoch 5/12
20000/20000 [==============================] - 184s 9ms/step - loss: 0.6932 - acc: 0.5010 - val_loss: 0.6935 - val_acc: 0.4938
Epoch 6/12
20000/20000 [==============================] - 184s 9ms/step - loss: 0.6932 - acc: 0.4995 - val_loss: 0.6936 - val_acc: 0.4938
Epoch 7/12
20000/20000 [==============================] - 184s 9ms/step - loss: 0.6932 - acc: 0.4990 - val_loss: 0.6932 -

In [17]:
from keras.preprocessing.sequence import pad_sequences
maxlen = 80
X = pad_sequences(dev_X, maxlen=maxlen)

In [23]:
model = Sequential([
    Embedding(NUM_WORDS, 128),
    LSTM(128, dropout=0.2, recurrent_dropout=0.2),
    Dense(1, activation="sigmoid")
])

In [24]:
model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [25]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_3 (Embedding)      (None, None, 128)         1280000   
_________________________________________________________________
lstm_3 (LSTM)                (None, 128)               131584    
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 129       
Total params: 1,411,713
Trainable params: 1,411,713
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model.fit(X, y, epochs=64, validation_split=0.2, batch_size=32)

Train on 20000 samples, validate on 5000 samples
Epoch 1/64
20000/20000 [==============================] - 50s 3ms/step - loss: 0.4731 - acc: 0.7755 - val_loss: 0.4152 - val_acc: 0.8090
Epoch 2/64
20000/20000 [==============================] - 50s 2ms/step - loss: 0.3334 - acc: 0.8596 - val_loss: 0.3850 - val_acc: 0.8284
Epoch 3/64
20000/20000 [==============================] - 50s 2ms/step - loss: 0.2536 - acc: 0.9002 - val_loss: 0.4171 - val_acc: 0.8340
Epoch 4/64
20000/20000 [==============================] - 50s 2ms/step - loss: 0.1968 - acc: 0.9257 - val_loss: 0.4622 - val_acc: 0.8274
Epoch 5/64
20000/20000 [==============================] - 50s 2ms/step - loss: 0.1550 - acc: 0.9417 - val_loss: 0.5350 - val_acc: 0.8170
Epoch 6/64
20000/20000 [==============================] - 50s 2ms/step - loss: 0.1284 - acc: 0.9521 - val_loss: 0.5556 - val_acc: 0.8184
Epoch 7/64
20000/20000 [==============================] - 50s 2ms/step - loss: 0.1004 - acc: 0.9638 - val_loss: 0.6361 - val_acc: